### PR TITLE
net: route: multicast routing feature

### PR DESCRIPTION
**Pull request by JanGeorgi**
net: route: Add prefix-based ipv6 multicast forwarding

This adds/reenables the feature of multicast routing/forwarding.
The forwarding decision is based on the added multicast routes
and the new network interface flag:
NET_IF_FORWARD_MULTICASTS.

Signed-off-by: Jan Georgi <jan.georgi@lemonbeat.com>

### CODE REVIEW

File: subsys/net/ip/Kconfig [hunk 1]
```diff
@@ -265,8 +265,11 @@ config NET_MAX_NEXTHOPS
 	  This determines how many entries can be stored in nexthop table.
 
 config NET_ROUTE_MCAST
-	bool
+	bool "Enable Multicast Routing / Forwarding"
+	default n
```
Comment (jukkar) [inline subsys/net/ip/Kconfig:269]: The option is "off" by default so no need to mention it, you can just remove the default line.

File: subsys/net/ip/ipv6.c [hunk 2]
```diff
@@ -362,6 +362,28 @@ static inline enum net_verdict ipv6_route_packet(struct net_pkt *pkt,
...
+#if defined(CONFIG_NET_ROUTE_MCAST)
+	/* check if routing loop could be created or if the destination is of
+	 * interface local scope or if from link local source
+	 */
+	if (net_ipv6_is_addr_mcast(&hdr->src)  ||
+	      net_ipv6_is_addr_mcast_iface(&hdr->dst) ||
+	       net_ipv6_is_ll_addr(&hdr->src)) {
+		return NET_CONTINUE;
+	}
+	int routed = net_route_mcast_forward_packet(pkt, hdr);
```
Comment (jukkar) [inline subsys/net/ip/ipv6.c:378]: Please declare the variable at the start of the function. Also please empty line after the }

File: subsys/net/ip/route.c [hunk 3]
```diff
@@ -644,6 +644,42 @@ int net_route_foreach(net_route_cb_t cb, void *user_data)
...
+				   const struct net_ipv6_hdr *hdr)
+{
+	int i, ret = 0, err = 0;
+
+	for (i = 0; i < CONFIG_NET_MAX_MCAST_ROUTES; ++i) {
+		struct net_route_entry_mcast *route = &route_mcast_entries[i];
+
+		if (!route->is_used) {
+			continue;
+		}
```
Comment (jukkar) [inline subsys/net/ip/route.c:657]: please place empty line after }

File: subsys/net/ip/route.c [hunk 4]
```diff
@@ -644,6 +644,42 @@ int net_route_foreach(net_route_cb_t cb, void *user_data)
...
+		if (!net_if_flag_is_set(route->iface,
+					NET_IF_FORWARD_MULTICASTS) ||
+		    !net_ipv6_is_prefix(hdr->dst.s6_addr,
+					route->group.s6_addr,
+					route->prefix_len)         ||
+		    (pkt->iface == route->iface)) {
+			continue;
+		}
+		struct net_pkt *pkt_cpy = net_pkt_shallow_clone(pkt, K_NO_WAIT);
```
Comment (jukkar) [inline subsys/net/ip/route.c:666]: Empty line after }
Also declare variable at the start of the block

File: subsys/net/ip/route.c [hunk 5]
```diff
@@ -644,6 +644,42 @@ int net_route_foreach(net_route_cb_t cb, void *user_data)
...
+
+		if (pkt_cpy == NULL) {
+			err--;
+			continue;
+		}
+		net_pkt_set_forwarding(pkt_cpy, true);
```
Comment (jukkar) [inline subsys/net/ip/route.c:672]: empty line

File: subsys/net/ip/route.c [hunk 6]
```diff
@@ -644,6 +644,42 @@ int net_route_foreach(net_route_cb_t cb, void *user_data)
...
+		net_pkt_set_iface(pkt_cpy, route->iface);
+		if (net_send_data(pkt_cpy) >= 0) {
+			++ret;
+		} else {
+			--err;
+		}
+	}
+	return (err == 0) ? ret : err;
```
Comment (jukkar) [inline subsys/net/ip/route.c:680]: same here

File: subsys/net/ip/route.c [hunk 7]
```diff
@@ -668,16 +706,24 @@ int net_route_mcast_foreach(net_route_mcast_cb_t cb,
...
-						  struct in6_addr *group)
+						  struct in6_addr *group,
+						  uint8_t prefix_len)
 {
 	int i;
 
+	if ((!net_if_flag_is_set(iface, NET_IF_FORWARD_MULTICASTS)) ||
+			(!net_ipv6_is_addr_mcast(group)) ||
+			(net_ipv6_is_addr_mcast_iface(group)) ||
+			(net_ipv6_is_addr_mcast_scope(group, 0x02))) {
```
Comment (jukkar) [inline subsys/net/ip/route.c:717]: Could you create a function `net_ipv6_is_addr_mcast_link()` in net_ip.h for this (similar what we have for other scope checks)?

File: subsys/net/ip/route.c [hunk 8]
```diff
@@ -668,16 +706,24 @@ int net_route_mcast_foreach(net_route_mcast_cb_t cb,
...
+		return NULL;
+	}
```
Comment (jukkar) [inline subsys/net/ip/route.c:719]: empty line after }

File: subsys/net/ip/route.c [hunk 9]
```diff
@@ -713,9 +759,12 @@ net_route_mcast_lookup(struct in6_addr *group)
 		struct net_route_entry_mcast *route = &route_mcast_entries[i];
 
 		if (!route->is_used) {
-			if (net_ipv6_addr_cmp(group, &route->group)) {
-				return route;
-			}
+			continue;
+		}
```
Comment (jukkar) [inline subsys/net/ip/route.c:763]: same here

File: subsys/net/ip/route.c [hunk 10]
```diff
@@ -713,9 +759,12 @@ net_route_mcast_lookup(struct in6_addr *group)
...
+		if (net_ipv6_is_prefix(group->s6_addr,
+				route->group.s6_addr,
```
Comment (jukkar) [inline subsys/net/ip/route.c:765]: Indendation seems bit off, can you verify that the parameters are aligned properly

File: tests/net/route_mcast/src/main.c [hunk 11]
```diff
@@ -0,0 +1,665 @@
...
+			&ipv6_hdr->src,
+			sizeof(struct in6_addr)) != 0) ||
+			(memcmp(&active_scenario.mcast,
+					&ipv6_hdr->dst,
+					sizeof(struct in6_addr)) != 0)) {
+		return false;
+	}
+
+	return true;
+}
```
Comment (jukkar) [inline tests/net/route_mcast/src/main.c:223]: empty line after }

File: tests/net/route_mcast/src/main.c [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+static struct dummy_api net_route_mcast_if_api_3 = {
+	.iface_api.init = net_route_mcast_iface_init3,
+	.send = iface_send,
+};
+
+#define _ETH_L2_LAYER DUMMY_L2
+#define _ETH_L2_CTX_TYPE NET_L2_GET_CTX_TYPE(DUMMY_L2)
+
+NET_DEVICE_INIT_INSTANCE(mcast_iface_1, "mcast_iface_1", iface_1,
+		net_route_mcast_dev_init, device_pm_control_nop,
```
Comment (jukkar) [inline tests/net/route_mcast/src/main.c:269]: indentation

File: tests/net/route_mcast/src/main.c [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+			 &net_route_data_if1, NULL,
+			 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+			 &net_route_mcast_if_api_1, _ETH_L2_LAYER,
+			 _ETH_L2_CTX_TYPE, 127);
+
+NET_DEVICE_INIT_INSTANCE(mcast_iface_2, "mcast_iface_2", iface_2,
+		net_route_mcast_dev_init, device_pm_control_nop,
```
Comment (jukkar) [inline tests/net/route_mcast/src/main.c:276]: same here

File: tests/net/route_mcast/src/main.c [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+			 &net_route_data_if2, NULL,
+			 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+			 &net_route_mcast_if_api_2, _ETH_L2_LAYER,
+			 _ETH_L2_CTX_TYPE, 127);
+
+NET_DEVICE_INIT_INSTANCE(mcast_iface_3, "mcast_iface_3", iface_3,
+		net_route_mcast_dev_init, device_pm_control_nop,
```
Comment (jukkar) [inline tests/net/route_mcast/src/main.c:283]: and here

File: tests/net/route_mcast/src/main.c [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+				      struct in6_addr *remote_addr,
+				      uint16_t src_port, uint16_t remote_port)
+{
+	static const char payload[] = "foobar";
+	struct net_pkt *pkt;
+	int res;
+
+	pkt = net_pkt_alloc_with_buffer(iface, strlen(payload), AF_INET6,
+					IPPROTO_UDP, K_FOREVER);
+
```
Comment (jukkar) [inline tests/net/route_mcast/src/main.c:300]: this empty line can be removed, we usually do not have have empty line between setting a var and then checking its value

**Review by jukkar — CHANGES_REQUESTED** [review]
Thanks for the PR, the tests were very nice to have. Overall looks good, just minor nits.

File: tests/net/route_mcast/src/main.c (hunk 11 above)
Comment (JanGeorgi) [reply tests/net/route_mcast/src/main.c:223]: Done. Not hit by automated check cause after the marked line.

File: subsys/net/ip/route.c (hunk 8 above)
Comment (JanGeorgi) [reply subsys/net/ip/route.c:725]: Done. Not hit by automated check cause after the marked line.

File: subsys/net/ip/route.c (hunk 7 above)
Comment (JanGeorgi) [reply subsys/net/ip/route.c]: Done.

File: subsys/net/ip/route.c (hunk 6 above)
Comment (JanGeorgi) [reply subsys/net/ip/route.c:686]: Done.

File: subsys/net/ip/route.c (hunk 5 above)
Comment (JanGeorgi) [reply subsys/net/ip/route.c:676]: Done.

File: subsys/net/ip/route.c (hunk 9 above)
Comment (JanGeorgi) [reply subsys/net/ip/route.c:770]: Done.

File: subsys/net/ip/route.c [hunk 16]
```diff
@@ -713,9 +766,13 @@ net_route_mcast_lookup(struct in6_addr *group)
...
 
 		if (!route->is_used) {
-			if (net_ipv6_addr_cmp(group, &route->group)) {
-				return route;
-			}
+			continue;
+		}
+
+		if (net_ipv6_is_prefix(group->s6_addr,
+							route->group.s6_addr,
```
Comment (jukkar) [inline subsys/net/ip/route.c:773]: There is indentation issue here. Do you have tab size set to 8 spaces, that might explain the indentation problems.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by chrta — APPROVED** [review]
Since i only took part in the design and during review, i think it is ok to approve here, too.

**Pull request merged**